### PR TITLE
Anomaly cleanups

### LINKED ISSUE #2146: Anomaly tidbits
small cherry-picks for Anomaly from PRs I intend to close..

### PR DESCRIPTION
**Pull request by breznak**
Fixes #2146

### CODE REVIEW

**Comment by breznak** [comment]
please review..should be simple cleanups/moving classes around

**Comment by rhyolight** [comment]
Please review @scottpurdy

File: nupic/algorithms/anomaly.py [hunk 1]
```diff
@@ -72,10 +72,11 @@ class Anomaly(object):
   _supportedModes = (MODE_PURE, MODE_LIKELIHOOD, MODE_WEIGHTED)
 
 
-  def __init__(self, slidingWindowSize=None, mode=MODE_PURE, binaryAnomalyThreshold=None):
+  def __init__(self, slidingWindowSize=0, mode=MODE_PURE, binaryAnomalyThreshold=0.0):
```
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: Isn't `None` a better default for disabled?

File: nupic/algorithms/anomaly.py [hunk 2]
```diff
@@ -87,26 +88,27 @@ def __init__(self, slidingWindowSize=None, mode=MODE_PURE, binaryAnomalyThreshol
...
     """
     self._mode = mode
-    if slidingWindowSize is not None:
+    if slidingWindowSize > 0:
       self._movingAverage = MovingAverage(windowSize=slidingWindowSize)
     else:
       self._movingAverage = None
 
-    if self._mode == Anomaly.MODE_LIKELIHOOD or self._mode == Anomaly.MODE_WEIGHTED:
+    if self._mode == Anomaly.MODE_LIKELIHOOD or \
```
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: wrap the entire condition in parens instead of using a backslash:
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: nupic/algorithms/anomaly.py [hunk 3]
```diff
@@ -87,26 +88,27 @@ def __init__(self, slidingWindowSize=None, mode=MODE_PURE, binaryAnomalyThreshol
...
                        "Anomaly.MODE_PURE, Anomaly.MODE_LIKELIHOOD, "
                        "Anomaly.MODE_WEIGHTED; you used: %r" % self._mode)
     self._binaryThreshold = binaryAnomalyThreshold
-    if binaryAnomalyThreshold is not None and ( 
-          not isinstance(binaryAnomalyThreshold, float) or
-          binaryAnomalyThreshold >= 1.0  or 
-          binaryAnomalyThreshold <= 0.0 ):
-      raise ValueError("Anomaly: binaryAnomalyThreshold must be from (0,1) "
-                       "or None if disabled.")
+    if not isinstance(binaryAnomalyThreshold, float) or \
```
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: I agree old was formatted oddly but stick with implicit line continuations inside parens

File: nupic/algorithms/anomaly.py [hunk 4]
```diff
@@ -130,15 +132,15 @@ def compute(self, activeColumns, predictedColumns,
     if self._mode == Anomaly.MODE_PURE:
       score = anomalyScore
     elif self._mode == Anomaly.MODE_LIKELIHOOD:
-      if inputValue is None:
-        raise ValueError("Selected anomaly mode 'Anomaly.MODE_LIKELIHOOD' "
-                 "requires 'inputValue' as parameter to compute() method. ")
-
+      assert inputValue is not None, ("Selected anomaly mode "
```
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: Don't use asserts when you really want to raise an exception. Assert statements are discarded when you run code in "optimized" mode so they are only appropriate for invariance checks that would be considered bugs if they were ever true.

File: nupic/algorithms/anomaly.py [hunk 5]
```diff
@@ -164,6 +166,17 @@ def __str__(self):
     return "Anomaly:\tmode=%s\twindowSize=%r" % (self._mode, windowSize)
 
 
+  def __cmp__(self, other):
```
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: What's the usage for this? I can imagine a use case for `__eq__` but `__cmp__` doesn't seem right.

File: nupic/utils.py [hunk 6]
```diff
@@ -97,6 +97,17 @@ def __setstate__(self, state):
       self.slidingWindow = sum(self.slidingWindow)
 
 
+  def __cmp__(self, o):
```
Comment (scottpurdy) [inline nupic/utils.py]: Again, this doesn't seem right.

File: nupic/utils.py [hunk 7]
```diff
@@ -114,3 +125,48 @@ def write(self, proto):
     proto.windowSize = self.windowSize
     proto.slidingWindow = self.slidingWindow
     proto.total = self.total
+
+###########################################################################
+class CircularBuffer(object):
+  """
+  implementation of a fixed size constant random access circular buffer
+  #TODO can this be merged with MovingAverage somehow?
```
Comment (scottpurdy) [inline nupic/utils.py]: Let's move this to a separate PR. The implementation of this class makes it look totally useless to me so I must not be understanding it correctly. We should take a look at the usage and probably replace with a regular Python list or `collections.deque`

File: nupic/algorithms/anomaly.py (hunk 1 above)
Comment (breznak) [inline nupic/algorithms/anomaly.py]: In one case yes, but for swarming, it makes it much nicer if you can go from `0..N`

File: nupic/utils.py (hunk 7 above)
Comment (breznak) [inline nupic/utils.py]: I think this was the intention behind it: https://github.com/numenta/nupic/pull/2149#discussion-diff-31208896R132
But I agree it might be useless actually...this is just moving it from inner class somewhere to light in utils.py. An issue to investigate usefulness?

**Comment by breznak** [comment]
Thanks guys for the review(s)! 
@rhyolight  I'll be AFK about a week, so I'll reply later on..

File: nupic/utils.py [hunk 8]
```diff
@@ -26,7 +26,7 @@
 
 import numbers
 
-
+###################################################################################
```
Comment (david-ragazzi) [inline nupic/utils.py]: Please remove this smudged line.

File: nupic/utils.py [hunk 9]
```diff
@@ -114,3 +125,48 @@ def write(self, proto):
     proto.windowSize = self.windowSize
     proto.slidingWindow = self.slidingWindow
     proto.total = self.total
+
+###########################################################################
```
Comment (david-ragazzi) [inline nupic/utils.py]: Please remove this smudged line.

File: nupic/algorithms/anomaly.py (hunk 1 above)
Comment (scottpurdy) [inline nupic/algorithms/anomaly.py]: But then you're forced to swarm over parameters from `0..N` rather than the range you want, which may be `20..50`. I'd rather change the permute variable to have a `tryNone` option which determines whether it tries `None` in addition to some integer/float range.

File: nupic/utils.py (hunk 7 above)
Comment (scottpurdy) [inline nupic/utils.py]: Yeah I don't know if we should keep it or not. I'd just rather not complicate this PR with an unrelated task of moving this class around.

**Comment by rhyolight** [comment]
@breznak Do you still plan to work on this?
